### PR TITLE
Issue #69: Introduce SearchEngineManager class providing access to a centralized registry of search engines.

### PR DESCRIPTION
**Pull request by pocmo**
Not 100% sure on the API yet. Will try to integrate this into Focus and see what's lacking.

### CODE REVIEW

File: components/browser/search/src/main/java/mozilla/components/browser/search/SearchEngineManager.kt [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+     * constructor of SearchEngineManager.
+     *
+     * Optionally an identifier can be passed to this method (e.g. from the user's preferences). If
+     * a matching search engine was loaded then this search engine will be returned instead.
+     */
+    @Synchronized
+    suspend fun getDefaultSearchEngine(context: Context, identifier: String = EMPTY): SearchEngine {
+        val searchEngines = getSearchEngines(context)
+
+        return when(identifier) {
```
Comment (sputnikci) [inline components/browser/search/src/main/java/mozilla/components/browser/search/SearchEngineManager.kt]: \[ktlint\] WARNING: \[keyword\-spacing\] Missing spacing after "when" in column 20

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 2]
```diff
@@ -0,0 +1,153 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.search
+
+import android.app.Application
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import android.content.Context
+import android.content.Intent
+import android.graphics.Bitmap
+import android.net.Uri
+import kotlinx.coroutines.experimental.delay
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.never
+import org.mockito.Mockito.spy
+import org.mockito.Mockito.verify
+import org.mockito.Mockito.verifyNoMoreInteractions
+import org.mockito.Mockito.verifyZeroInteractions
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import org.robolectric.Robolectric
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.RuntimeEnvironment
+import org.robolectric.Shadows
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+import org.robolectric.shadows.ShadowApplication
+import org.robolectric.shadows.ShadowLooper
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[no\-unused\-imports\] Unused import in column 1

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+
+    private fun List<SearchEngine>.assertContainsIdentifier(identifier: String) {
+        if (find { it.identifier == identifier } == null) {
+            throw AssertionError("$identifier not in list")
+        }
+    }
+
+    private fun mockProvider(engines: List<SearchEngine>) : SearchEngineProvider =
+            object : SearchEngineProvider {
+                override suspend fun loadSearchEngines(context: Context): List<SearchEngine> {
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[modifier\-order\] Incorrect modifier order \(should be "suspend override"\) in column 17

File: components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+    private fun List<SearchEngine>.assertContainsIdentifier(identifier: String) {
+        if (find { it.identifier == identifier } == null) {
+            throw AssertionError("$identifier not in list")
+        }
+    }
+
+    private fun mockProvider(engines: List<SearchEngine>): SearchEngineProvider =
+            object : SearchEngineProvider {
+                override suspend fun loadSearchEngines(context: Context): List<SearchEngine> {
```
Comment (sputnikci) [inline components/browser/search/src/test/java/mozilla/components/browser/search/SearchEngineManagerTest.kt]: \[ktlint\] WARNING: \[modifier\-order\] Incorrect modifier order \(should be "suspend override"\) in column 17

**Comment by pocmo** [comment]
This is now ready and I have a branch ready to get this integrated into Focus. 🎉

**Comment by csadilek** [comment]
@pocmo looks great to me! I think we can merge and release :).

**Pull request merged**